src/wallet/rollover.rs

```diff
@@ -0,0 +1,453 @@
+            Enc_n_prime_B_1,
+            P,
+            V,
+            Com_w
+        ),
+        (
+            B,
+            B_blinding,
+            X_1
+        )
+        :
+        Enc_n_prime_B_0 = (r_n * B),
+        Enc_n_prime_B_1 = (n_prime * B + r_n * D),
+        Enc_w_B_0 = (r_w * B),
+        Enc_w_B_1 = (w * B + r_w * D),
+        Com_w = (w * P + w_blinding * B_blinding),
+        V = (w_blinding * X_1 + minus_r_Q * B)
+    }
+
+    define_proof! {
+        issuer,
+        "WalletRollover_Issuer",
+        (
+            b,
+            r,
+            x_0,
+            x_1,
+            x_2,
+            x_0_blinding,
+            x_prime_0,
+            x_prime_1,
+            x_prime_2,
+            x_prime_0_blinding,
+            t_1,
+            t_2
+        ),
+        (
+            P,
+            D,
+            Enc_w_B_0,
+            Enc_w_B_1,
+            Enc_n_prime_B_0,
+            Enc_n_prime_B_1,
+            Enc_Q_0,
+            Enc_Q_1,
+            T_1_a,
+            T_1_b,
+            T_2_a,
+            T_2_b
+        ),
+        (
+            X_0,
+            X_1,
+            X_2,
+            X_prime_0,
+            X_prime_1,
+            X_prime_2,
+            B,
+            B_blinding
+        )
+        :
+        X_0 = (x_0 * B + x_0_blinding * B_blinding),
+        X_1 = (x_1 * B_blinding),
+        X_2 = (x_2 * B_blinding),
+        X_prime_0 = (x_prime_0 * B + x_prime_0_blinding * B_blinding),
+        X_prime_1 = (x_prime_1 * B_blinding),
+        X_prime_2 = (x_prime_2 * B_blinding),
+        P = (b * B),
+        T_1_a = (b * X_prime_1),
+        T_1_b = (t_1 * B_blinding),
+        T_2_a = (b * X_prime_2),
+        T_2_b = (t_2 * B_blinding),
+        Enc_Q_0 = (r * B + t_1 * Enc_w_B_0 + t_2 * Enc_n_prime_B_0),
+        Enc_Q_1 = (x_prime_0 * P + r * D + t_1 * Enc_w_B_1 + t_2 * Enc_n_prime_B_1)
+    }
+}
+
+/// A request for wallet rollover.
+#[allow(non_snake_case)]
+pub struct Request {
+    epoch: Epoch,
+    new_epoch: Epoch,
+    n: Scalar,
+    D: CompressedRistretto,
+    Enc_n_prime_B: (CompressedRistretto, CompressedRistretto),
+    Enc_w_B: (CompressedRistretto, CompressedRistretto),
+    Com_w: CompressedRistretto,
+    P: CompressedRistretto,
+    C_Q: CompressedRistretto,
+    proof: proofs::client::CompactProof,
+}
+
+/// State held by the client while awaiting a wallet rollover response.
+#[allow(non_snake_case)]
+pub struct AwaitingResponse {
+    old_parameters: Parameters,
+    new_parameters: Parameters,
+    transcript: Transcript,
+    n_prime: Scalar,
+    w: u64,
+    d: Scalar,
+    D: RistrettoPoint,
+    Enc_w_B: (CompressedRistretto, CompressedRistretto),
+    Enc_n_prime_B: (CompressedRistretto, CompressedRistretto),
+}
+
+impl Wallet {
+    #[allow(non_snake_case)]
+    pub fn request_rollover<R: RngCore + CryptoRng>(
+        self,
+        old_parameters: &Parameters,
+        new_parameters: &Parameters,
+        mut transcript: Transcript,
```

Comment:
Just noting this for later: in a future issue, we may want to consider the misuse-resistance tradeoffs of not taking a `Transcript`, to prevent users from reusing `Transcript`s across calls that are supposed to be anonymous.

Reply:
https://github.com/hdevalence/danake/issues/20